src/transport/manager/handle.rs

```diff
@@ -180,51 +180,67 @@ impl TransportManagerHandle {
         peer: &PeerId,
         addresses: impl Iterator<Item = Multiaddr>,
     ) -> usize {
-        let mut peers = self.peers.write();
-        let addresses = addresses
-            .filter_map(|address| {
-                (self.supported_transport(&address) && !self.is_local_address(&address))
-                    .then_some(AddressRecord::from_multiaddr(address)?)
-            })
-            .collect::<HashSet<_>>();
-
-        // if all of the added addresses belonged to unsupported transports, exit early
-        let num_added = addresses.len();
-        if num_added == 0 {
-            tracing::debug!(
-                target: LOG_TARGET,
-                ?peer,
-                "didn't add any addresses for peer because transport is not supported",
-            );
+        let mut peer_addresses = HashSet::new();
 
-            return 0usize;
+        for address in addresses {
+            // There is not supported transport configured that can dial this address.
+            if !self.supported_transport(&address) {
+                continue;
+            }
+            if self.is_local_address(&address) {
+                continue;
+            }
+
+            // Check the peer ID if present.
+            if let Some(Protocol::P2p(multihash)) = address.iter().last() {
+                // Ignore the address if the peer ID is invalid.
+                let Ok(peer_id) = PeerId::from_multihash(multihash.clone()) else {
+                    continue;
+                };
+
+                // This can correspond to the provided peerID or to a different one.
+                if peer_id != *peer {
```

Comment:
nit: We can rewrite this as `multihash != peer.as_ref()` without the need to upcast the multihash above.